### ISSUE TITLE
~/.cache/crystal vs. Linux distro updates

### ISSUE THREAD

**Issue by petr-fischer** [crystal-lang/crystal#9314]
Hello! When I updated Ubuntu from 19.04 to the 20.04, there is new `libevent` library version.

After that, my compilations with Crystal 0.34 started to break (linker errors) until I deleted `~/.cache/crystal`.

I assume that in Crystal binary caches (in `~/.cache/crystal`) there are fixed symbols to the libraries (like libevent, etc.), so if you update Linux distro with new versions of this libraries, the compiler tries to use these caches, which point to a library that no longer exists and -> bang, linker errors.

I understand, that this is not a direct Crystal problem (maybe), but for the first time, this was super weird for me (and for my colleagues too). Is there a way the compiler can cope better or advise developers better? I believe this can be a big problem for some people and they will start randomly reinstalling everything possible in this situation.

**Comment by j8r** [comment]
The compiler could try first to link the project, as now. Then, issue a warning on failure, delete the project cache in `~/.crystal/cache` before trying again automatically?

It sounds good to do this, because most of linking failures are:
- missing libraries, not installed on the system, required for a new project
- corrupted/invalid cache

It is a common issue, most of us has faced. The cache first purpose is to speed up compilation, too bad it can breaks compilation.

**Comment by RX14** [comment]
I think clearing the cache on a link failure is a good idea.

**Comment by jhass** [comment]
I actually had the same idea, if the cache was non-empty at invocation, let's clear the cache, print a warning and completely reboot compilation once on link failure and only then error out if it fails the second time.

**Comment by naqvis** [comment]
or for time-being (until this is fixed) at least in docs or reference material make it **obvious** that its a good practice to clean cache prior to compilation?

`rm -rf $(crystal env CRYSTAL_CACHE_DIR)`

**Comment by waj** [comment]
I don't think the compiler has to do any action when it fails to compile. It may obscure things and make diagnosis more complicated. At most it could suggest try cleaning the cache when it fails at the linking stage.

**Comment by jhass** [comment]
As a compromise we could also make this interactive if stdin is a tty.

```
<linker error output>
Linking failed! The cache was non-empty at the start of the compilation. Clear cache and retry compilation? [Y/n]
Successfully built <foo.cr>
```
If stdin is non-interactive just abort like now.

**Comment by petr-fischer** [comment]
@waj - don't forget that it's an obsolete cache problem - and - who is responsible for that cache? the compiler, of course :)
But some clear suggestion or better message to the developer is probably sufficient (IMHO).

**Comment by naqvis** [comment]
> I don't think the compiler has to do any action when it fails to compile.

I would argue against it, cache folder is created by the compiler, so it should also be responsibility of it to manage it. we shouldn't be asking users to manage  things which they didn't initiated or have no control of.  imho asking user to manually delete cache is just kind of temporary workaround, but not a permanent solution.

**Comment by waj** [comment]
If the cache makes the compiler to fail then it's clearly a bug. We might have to admit it's probably a bug hardly feasible to fix 100%, so a message with a suggestion about a possible action sounds reasonable.

Making assumptions, re-trying compilation or making the compiler interactive doesn't sound like the right approach to me. I wouldn't make the compiler even more complicated for this edge case.

**Comment by straight-shoota** [comment]
If the compiler automatically clears the cache, it would be incredibly hard to debug cache issues.

It sounds like a good idea to print a message suggesting to clear the cache on linking errors.

**Comment by Blacksmoke16** [comment]
Worth abstracting the command to clear the cache?  Something like `crystal clear-cache`.  This way the user doesn't have to know where it is.  Plus the location could internally change yet the API would be the same.

**Comment by naqvis** [comment]
`crystal env` does provide the location of cache folder.

`rm -rf $(crystal env CRYSTAL_CACHE_DIR)`

**Comment by waj** [comment]
@Blacksmoke16 that sounds good to me. Maybe `crystal cache --clear` subcommand would enable other options later.
But it's true as @naqvis said, `crystal env` already provides the location of the cache.

**Comment by Blacksmoke16** [comment]
> `crystal env` does provide the location of cache folder.

Oh, neat.  First time I ever ran that 😆.

> Maybe `crystal cache --clear` subcommand would enable other options later.

Yea sure, there are options.  I'm not tied to a specific name/command, just the general ideal.

**Comment by jhass** [comment]
Maybe a `--clean-cache` or `--no-cache` option to `crystal build`? Then the error  could just say ".. Try rerunning with --clean-cache`.

**Comment by HankAnderson** [comment]
I disagree. The users shouldn't even have to know about the cache. Things should just work. If the users need to clear the cache, it's because of a bug in the compiler. Once the bug is fixed there won't be a need to clear the cache, so that command is useless.

**Comment by straight-shoota** [comment]
I'd rather just have the compiler directly print the command.
```
Try running `rm -rf ~/.crystal/cache` to clear the compiler cache
```
The path should obviously be the same value as `crystal env CRYSTAL_CACHE_DIR`. But there's no need for redirection here. The cache dir won't change between the link failure and running `rm`.

**Comment by RX14** [comment]
> If the compiler automatically clears the cache, it would be incredibly hard to debug cache issues.

There could be an env var or option to turn this off. We should weigh up making debugging harder over making lives easier for those why don't know to clear the cache. I think both situations are honestly rare (this is certainly not a common issues) so it's fairly hard to toss up.

**Comment by straight-shoota** [comment]
> making lives easier for those why don't know to clear the cache

What's hard about following the instruction?
```
Try running `rm -rf ~/.crystal/cache` to clear the compiler cache.
```

**Comment by RX14** [comment]
It might be better, not sure. I don't mind.

**Comment by j8r** [comment]
The user should not have to clean manually the cache to make the crystal compiler happy.

If the compiler can't use the cache it has created, this is its job to fix it.
The compiler can remove its cache, then a dedicated clean command won't even be necessary.

**Comment by j8r** [comment]
I would like to highlight the fact that the cache is **optional**. We can perfectly compile crystal applications without, not a big deal. It is just slower, that's why there will be a warning saying that the cache can't be used.

**Comment by waj** [comment]
I think we're diverging from the original issue here (that includes me). If anyone has some feature to propose, please open a RFC or send a PR.

We should focus on the original issue instead, because I still don't understand how this could be related with the upgrade. AFAIK the .o files doesn't have any reference to the previous library, just symbol names. Can someone provide some insight here?

**Comment by petr-fischer** [comment]
@waj - OK, as the creator of this issue, I will try the whole workflow again (create Crystal/Kemal project on the Ubuntu 19.04 VirtualBox machine, compile it [create caches], then upgrade to the 20.04 and compile again). Wait few hours please...

Who bets I won't run into this problem again? :) (but my colleague confirmed this behavior to me)

**Comment by j8r** [comment]
@waj Not really, it is not hard to corrupt the cache by CTRL+C near the end of the compilation of our program (easier bigger the project is). True I am distracted and impatient at times :smile: 

Indeed it will be great to solve this issue in particular, but I think the issue is broader. Not all cache corruptions can be solved. I am just saying:
cache is not mandatory - if this feature does not work for some reasons, let's not bother with is it, warn, remove and try again without.

**Comment by petr-fischer** [comment]
Of course, I am not able to simulate this issue in a virtual machine with a "few lines" Crystal/Kemal project. Probably my previous big ~/.cache/crystal caches on the dev laptop was crippled in some other way :(

**Comment by bcardiff** [comment]
I like `--clean-cache` / `--no-cache` proposed in https://github.com/crystal-lang/crystal/issues/9314#issuecomment-630870311 since it hide the knowledge of how to discover the crystal cache path which can vary from between setups.

And via `CRYSTAL_OPTS`, it can be set as expected in https://github.com/crystal-lang/crystal/issues/9314#issuecomment-630908797

**Comment by yxhuvud** [comment]
It hasn't been mentioned in this thread, but how about changing `~/.cache/crystal` to `~/.cache/crystal/$CRYSTAL_VERSION`? Would solve the issue at hand for most users, as well as simplifying having multiple crystals installed.. 

Downside being that old compilation caches would hang around forever. Dunno how big an issue that would be.

**Comment by straight-shoota** [comment]
For a most users who stumbles onto this, it probably doesn't make any difference whether the compiler suggests to rerun the compiler command with `--clean-cache` or run `rm ~/.cache/crystal` (with appropriate path for the specific setup).
The latter is easy and should always work. But the `--clean-cache` flag might not be easy to apply if the user doesn't invoke the compiler directly. `CRYSTAL_OPTS` would probably work, but it's an additional step.

IMO `rm ~/.cache/crystal` is the best and easiest solution.
`--clean-cache` adds more complexity to the compiler (for an uncommon case) and makes it more difficult to use.
Are there any problems speaking against `rm ~/.cache/crystal`?

@j8r  We can certainly try to improve cache robusness, but a blunt automatic "clear cache and try again" behaviour is a bad idea. It even prevents further improvements when issues don't even surface anymore. Instructing the user to run a simple command to clear the cache is a better solution for this edge case problem.

**Comment by j8r** [comment]
@straight-shoota it could just be a default behavior, with an option to fail when the cache is corrupted

Most users will just do what the compiler tell them to do, it won't change much to have to do it manually vs automatically done by the compiler.

Also, the error message will still be printed anyway, like now.

**Comment by RX14** [comment]
Thinking about it more, I actually agree with @waj. The cache is from the LLVM IR to a `.o` file. If the library symbols we use change, the IR changes too. We need to recreate this bug so we can work out the root cause of this, because this bug doesn't quite add up as-is.

Once we know the root cause, we can then go about fixing it in the most appropriate way.

**Comment by asterite** [comment]
Steps to reproduce:

```bash
$ rm -rf $(crystal env CRYSTAL_CACHE_DIR)
$ echo 'puts "hello world"' > foo.cr
$ crystal build foo.cr
$ cd $(crystal env CRYSTAL_CACHE_DIR)
```

Run `ls` there and `cd` into the first directory.

Then run this:

```bash
$ cp U-I-nt64.o _main.o
```

Then go back to where `foo.cr` is defined and compile again:

```bash
$ crystal build foo.cr
```

You get a linker error saying a lot of symbols are missing, which makes sense because I just overwrote an `.o` file with another `.o` file.

I think in this case it's harmless if the compiler tries to erase the cache dir for this particular file and try again. If there's a linker error, well, it's a legit linker error (probably some C binding that isn't fulfilled). But if it was because of a stale `.o` file, or become someone intentionally tried to break things, then it will work fine.

I'll try to submit a PR for this.

**Comment by petr-fischer** [comment]
Another cache issue from the forums (it's maybe still the same issue, I do not have time right now), just for the reference, FYI:
https://forum.crystal-lang.org/t/compilation-error-in-crystal-0-35-1/2283
